File: Data Structure/6198번 - 옥상 정원 꾸미기/6198.cpp
#include <iostream>
#include <vector>
#include <stack>

using namespace std;

int n;
vector<int> v;

void input() {
    cin >> n;
    v.resize(n + 1);
    for(int i = 0; i < n; i++) {
        cin >> v[i];
    }
}

void solve() {
    stack<int> roof;
    long long ans = 0;
    
    for(int i = 0; i < n; i++) {
        while(!roof.empty() && roof.top() <= v[i]) {
            roof.pop();
        }
        ans += roof.size();
        roof.push(v[i]);
    }
    
    cout << ans << '\n';
}

int main() {
    ios::sync_with_stdio(0);
    cin.tie(0);
    
    input();
    solve();
    
    return 0;
}

File: Data Structure/6198번 - 옥상 정원 꾸미기/6198.py
import sys

input = sys.stdin.readline

n = int(input())
arr = list(int(input()) for _ in range(n))

def solve():
    roof = []
    ans = 0
    
    for el in arr:
        while len(roof) and roof[-1] <= el:
            roof.pop()
        ans += len(roof)
        roof.append(el)
    
    print(ans)

if __name__ == '__main__':
    solve()